src/dlfcn.c

```diff
@@ -467,7 +465,162 @@ char *dlerror( void )
     return error_buffer;
 }
 
-#ifdef SHARED
+/* taken from http://bandido.ch/programming/Import_Address_Table_Hooking.pdf */
+static IMAGE_IMPORT_DESCRIPTOR* getImportTable( HMODULE module, DWORD *size )
+{
+    IMAGE_DOS_HEADER* dosHeader = (IMAGE_DOS_HEADER*)module;
+    if ( dosHeader->e_magic != 0x5A4D )
+        return NULL;
+    IMAGE_OPTIONAL_HEADER* optionalHeader = (IMAGE_OPTIONAL_HEADER*)
+        ((BYTE*)module + dosHeader->e_lfanew + 24);
+    if ( optionalHeader->Magic != IMAGE_NT_OPTIONAL_HDR_MAGIC )
+        return NULL;
+    if ( optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size == 0 ||
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress == 0 )
+        return NULL;
+    *size = optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].Size;
+    return ( IMAGE_IMPORT_DESCRIPTOR*)((BYTE*)module +
+        optionalHeader->DataDirectory[IMAGE_DIRECTORY_ENTRY_IMPORT].VirtualAddress );
+}
+/*
+ * return symbol name for a given address
+ */
+static char *getSymbolName( HMODULE baseAddress, IMAGE_IMPORT_DESCRIPTOR *iid, void *addr )
+{
+    for( int i = 0; iid[i].Characteristics != 0 && iid[i].FirstThunk != 0; i++ ) {
+        PIMAGE_THUNK_DATA thunkILT = (PIMAGE_THUNK_DATA)( iid[i].Characteristics + baseAddress );
+        PIMAGE_THUNK_DATA thunkIAT = (PIMAGE_THUNK_DATA)( iid[i].FirstThunk + baseAddress );
+        for( ; thunkILT->u1.AddressOfData != 0; thunkILT++, thunkIAT++ ) {
+            if( IMAGE_SNAP_BY_ORDINAL(thunkILT->u1.Ordinal) )
+                continue;
+            if( (void *)thunkIAT->u1.Function != addr )
+                continue;
+            PIMAGE_IMPORT_BY_NAME nameData = (PIMAGE_IMPORT_BY_NAME)( thunkILT->u1.AddressOfData + baseAddress );
+            return nameData->Name;
+        }
+   }
+   return NULL;
+}
+
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static unsigned char *getAddressFromIAT( unsigned char *addr )
+{
+    BYTE *p = (unsigned char *)addr;
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( *p == 0xff && *(p+1) == 0x25 )
+    {

```

Comment:
Another simplification = removing one indentation: return error (NULL) as early as possible:
```c
if( addr[0] != 0xff || addr[1] != 0x25 )
    return NULL;
```
(and casting via `p` should not be needed)

Reply:
partially applied, the removal of p will be in a following commit.